// @nomicfoundation/ethereumjs-blockchain/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DBSetTD = exports.DBSetHashToNumber = exports.DBSetBlockOrHeader = exports.DBSaveLookups = exports.DBOp = exports.EthashConsensus = exports.CliqueConsensus = exports.CasperConsensus = exports.Blockchain = void 0;
var blockchain_js_1 = require("./blockchain.js");
Object.defineProperty(exports, "Blockchain", { enumerable: true, get: function () { return blockchain_js_1.Blockchain; } });
var index_js_1 = require("./consensus/index.js");
Object.defineProperty(exports, "CasperConsensus", { enumerable: true, get: function () { return index_js_1.CasperConsensus; } });
Object.defineProperty(exports, "CliqueConsensus", { enumerable: true, get: function () { return index_js_1.CliqueConsensus; } });
Object.defineProperty(exports, "EthashConsensus", { enumerable: true, get: function () { return index_js_1.EthashConsensus; } });
var helpers_js_1 = require("./db/helpers.js");
Object.defineProperty(exports, "DBOp", { enumerable: true, get: function () { return helpers_js_1.DBOp; } });
Object.defineProperty(exports, "DBSaveLookups", { enumerable: true, get: function () { return helpers_js_1.DBSaveLookups; } });
Object.defineProperty(exports, "DBSetBlockOrHeader", { enumerable: true, get: function () { return helpers_js_1.DBSetBlockOrHeader; } });
Object.defineProperty(exports, "DBSetHashToNumber", { enumerable: true, get: function () { return helpers_js_1.DBSetHashToNumber; } });
Object.defineProperty(exports, "DBSetTD", { enumerable: true, get: function () { return helpers_js_1.DBSetTD; } });
__exportStar(require("./types.js"), exports);
