// @nomicfoundation/ethereumjs-evm/dist/cjs/evm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultBlock = exports.EvmErrorResult = exports.CodesizeExceedsMaximumError = exports.INVALID_EOF_RESULT = exports.INVALID_BYTECODE_RESULT = exports.COOGResult = exports.OOGResult = exports.EVM = void 0;
const ethereumjs_common_1 = require("@nomicfoundation/ethereumjs-common");
const ethereumjs_statemanager_1 = require("@nomicfoundation/ethereumjs-statemanager");
const ethereumjs_util_1 = require("@nomicfoundation/ethereumjs-util");
const debug_1 = require("debug");
const eof_js_1 = require("./eof.js");
const exceptions_js_1 = require("./exceptions.js");
const interpreter_js_1 = require("./interpreter.js");
const journal_js_1 = require("./journal.js");
const logger_js_1 = require("./logger.js");
const message_js_1 = require("./message.js");
const index_js_1 = require("./opcodes/index.js");
const index_js_2 = require("./precompiles/index.js");
const transientStorage_js_1 = require("./transientStorage.js");
const types_js_1 = require("./types.js");
const { debug: createDebugLogger } = debug_1.default;
const debug = createDebugLogger('evm:evm');
const debugGas = createDebugLogger('evm:gas');
const debugPrecompiles = createDebugLogger('evm:precompiles');
/**
 * EVM is responsible for executing an EVM message fully
 * (including any nested calls and creates), processing the results
 * and storing them to state (or discarding changes in case of exceptions).
 * @ignore
 */
class EVM {
    constructor(opts = {}) {
        /**
         * EVM is run in DEBUG mode (default: false)
         * Taken from DEBUG environment variable
         *
         * Safeguards on debug() calls are added for
         * performance reasons to avoid string literal evaluation
         * @hidden
         */
        this.DEBUG = false;
        this.events = new ethereumjs_util_1.AsyncEventEmitter();
        this._optsCached = opts;
        this.transientStorage = new transientStorage_js_1.TransientStorage();
        if (opts.common) {
            this.common = opts.common;
        }
        else {
            const DEFAULT_CHAIN = ethereumjs_common_1.Chain.Mainnet;
            this.common = new ethereumjs_common_1.Common({ chain: DEFAULT_CHAIN });
        }
        let blockchain;
        if (opts.blockchain === undefined) {
            blockchain = new types_js_1.DefaultBlockchain();
        }
        else {
            blockchain = opts.blockchain;
        }
        this.blockchain = blockchain;
        this.stateManager = opts.stateManager ?? new ethereumjs_statemanager_1.DefaultStateManager();
        // Supported EIPs
        const supportedEIPs = [
            1153, 1559, 2315, 2565, 2718, 2929, 2930, 3074, 3198, 3529, 3540, 3541, 3607, 3651, 3670,
            3855, 3860, 4399, 4895, 4788, 4844, 5133, 5656, 6780, 6800, 7516,
        ];
        for (const eip of this.common.eips()) {
            if (!supportedEIPs.includes(eip)) {
                throw new Error(`EIP-${eip} is not supported by the EVM`);
            }
        }
        if (!EVM.supportedHardforks.includes(this.common.hardfork())) {
            throw new Error(`Hardfork ${this.common.hardfork()} not set as supported in supportedHardforks`);
        }
        this.allowUnlimitedContractSize = opts.allowUnlimitedContractSize ?? false;
        this.allowUnlimitedInitCodeSize = opts.allowUnlimitedInitCodeSize ?? false;
        this._customOpcodes = opts.customOpcodes;
        this._customPrecompiles = opts.customPrecompiles;
        this.journal = new journal_js_1.Journal(this.stateManager, this.common);
        this.common.events.on('hardforkChanged', () => {
            this.getActiveOpcodes();
            this._precompiles = (0, index_js_2.getActivePrecompiles)(this.common, this._customPrecompiles);
        });
        // Initialize the opcode data
        this.getActiveOpcodes();
        this._precompiles = (0, index_js_2.getActivePrecompiles)(this.common, this._customPrecompiles);
        this._emit = async (topic, data) => {
            return new Promise((resolve) => this.events.emit(topic, data, resolve));
        };
        this.performanceLogger = new logger_js_1.EVMPerformanceLogger();
        // Skip DEBUG calls unless 'ethjs' included in environmental DEBUG variables
        // Additional window check is to prevent vite browser bundling (and potentially other) to break
        this.DEBUG =
            typeof window === 'undefined' ? process?.env?.DEBUG?.includes('ethjs') ?? false : false;
    }
    get precompiles() {
        return this._precompiles;
    }
    get opcodes() {
        return this._opcodes;
    }
    /**
     * Returns a list with the currently activated opcodes
     * available for EVM execution
     */
    getActiveOpcodes() {
        const data = (0, index_js_1.getOpcodesForHF)(this.common, this._customOpcodes);
        this._opcodes = data.opcodes;
        this._dynamicGasHandlers = data.dynamicGasHandlers;
        this._handlers = data.handlers;
        this._opcodeMap = data.opcodeMap;
        return data.opcodes;
    }
    async _executeCall(message) {
        let account = await this.stateManager.getAccount(message.authcallOrigin ?? message.caller);
        if (!account) {
            account = new ethereumjs_util_1.Account();
        }
        let errorMessage;
        // Reduce tx value from sender
        if (!message.delegatecall) {
            try {
                await this._reduceSenderBalance(account, message);
            }
            catch (e) {
                errorMessage = e;
            }
        }
        // Load `to` account
        let toAccount = await this.stateManager.getAccount(message.to);
        if (!toAccount) {
            toAccount = new ethereumjs_util_1.Account();
        }
        // Add tx value to the `to` account
        if (!message.delegatecall) {
            try {
                await this._addToBalance(toAccount, message);
            }
            catch (e) {
                errorMessage = e;
            }
        }
        // Load code
        await this._loadCode(message);
        let exit = false;
        if (!message.code || (typeof message.code !== 'function' && message.code.length === 0)) {
            exit = true;
            if (this.DEBUG) {
                debug(`Exit early on no code (CALL)`);
            }
        }
        if (errorMessage !== undefined) {
            exit = true;
            if (this.DEBUG) {
                debug(`Exit early on value transfer overflowed (CALL)`);
            }
        }
        if (exit) {
            return {
                execResult: {
                    gasRefund: message.gasRefund,
                    executionGasUsed: ethereumjs_util_1.BIGINT_0,
                    exceptionError: errorMessage,
                    returnValue: new Uint8Array(0),
                },
            };
        }
        let result;
        if (message.isCompiled) {
            let timer;
            let target;
            if (this._optsCached.profiler?.enabled === true) {
                target = (0, ethereumjs_util_1.bytesToUnprefixedHex)(message.codeAddress.bytes);
                // TODO: map target precompile not to address, but to a name
                target = (0, index_js_2.getPrecompileName)(target) ?? target.slice(20);
                timer = this.performanceLogger.startTimer(target);
            }
            result = await this.runPrecompile(message.code, message.data, message.gasLimit);
            if (this._optsCached.profiler?.enabled === true) {
                this.performanceLogger.stopTimer(timer, Number(result.executionGasUsed), 'precompiles');
            }
            result.gasRefund = message.gasRefund;
        }
        else {
            if (this.DEBUG) {
                debug(`Start bytecode processing...`);
            }
            result = await this.runInterpreter(message);
        }
        if (message.depth === 0) {
            this.postMessageCleanup();
        }
        return {
            execResult: result,
        };
    }
    async _executeCreate(message) {
        let account = await this.stateManager.getAccount(message.caller);
        if (!account) {
            account = new ethereumjs_util_1.Account();
        }
        // Reduce tx value from sender
        await this._reduceSenderBalance(account, message);
        if (this.common.isActivatedEIP(3860)) {
            if (message.data.length > Number(this.common.param('vm', 'maxInitCodeSize')) &&
                !this.allowUnlimitedInitCodeSize) {
                return {
                    createdAddress: message.to,
                    execResult: {
                        returnValue: new Uint8Array(0),
                        exceptionError: new exceptions_js_1.EvmError(exceptions_js_1.ERROR.INITCODE_SIZE_VIOLATION),
                        executionGasUsed: message.gasLimit,
                    },
                };
            }
        }
        message.code = message.data;
        message.data = new Uint8Array(0);
        message.to = await this._generateAddress(message);
        if (this.common.isActivatedEIP(6780)) {
            message.createdAddresses.add(message.to.toString());
        }
        if (this.DEBUG) {
            debug(`Generated CREATE contract address ${message.to}`);
        }
        let toAccount = await this.stateManager.getAccount(message.to);
        if (!toAccount) {
            toAccount = new ethereumjs_util_1.Account();
        }
        // Check for collision
        if ((toAccount.nonce && toAccount.nonce > ethereumjs_util_1.BIGINT_0) ||
            !((0, ethereumjs_util_1.equalsBytes)(toAccount.codeHash, ethereumjs_util_1.KECCAK256_NULL) === true)) {
            if (this.DEBUG) {
                debug(`Returning on address collision`);
            }
            return {
                createdAddress: message.to,
                execResult: {
                    returnValue: new Uint8Array(0),
                    exceptionError: new exceptions_js_1.EvmError(exceptions_js_1.ERROR.CREATE_COLLISION),
                    executionGasUsed: message.gasLimit,
                },
            };
        }
        await this.journal.putAccount(message.to, toAccount);
        await this.stateManager.clearContractStorage(message.to);
        const newContractEvent = {
            address: message.to,
            code: message.code,
        };
        await this._emit('newContract', newContractEvent);
        toAccount = await this.stateManager.getAccount(message.to);
        if (!toAccount) {
            toAccount = new ethereumjs_util_1.Account();
        }
        // EIP-161 on account creation and CREATE execution
        if (this.common.gteHardfork(ethereumjs_common_1.Hardfork.SpuriousDragon)) {
            toAccount.nonce += ethereumjs_util_1.BIGINT_1;
        }
        // Add tx value to the `to` account
        let errorMessage;
        try {
            await this._addToBalance(toAccount, message);
        }
        catch (e) {
            errorMessage = e;
        }
        let exit = false;
        if (message.code === undefined ||
            (typeof message.code !== 'function' && message.code.length === 0)) {
            exit = true;
            if (this.DEBUG) {
                debug(`Exit early on no code (CREATE)`);
            }
        }
        if (errorMessage !== undefined) {
            exit = true;
            if (this.DEBUG) {
                debug(`Exit early on value transfer overflowed (CREATE)`);
            }
        }
        if (exit) {
            return {
                createdAddress: message.to,
                execResult: {
                    executionGasUsed: ethereumjs_util_1.BIGINT_0,
                    gasRefund: message.gasRefund,
                    exceptionError: errorMessage,
                    returnValue: new Uint8Array(0),
                },
            };
        }
        if (this.DEBUG) {
            debug(`Start bytecode processing...`);
        }
        let result = await this.runInterpreter(message);
        // fee for size of the return value
        let totalGas = result.executionGasUsed;
        let returnFee = ethereumjs_util_1.BIGINT_0;
        if (!result.exceptionError) {
            returnFee =
                BigInt(result.returnValue.length) * BigInt(this.common.param('gasPrices', 'createData'));
            totalGas = totalGas + returnFee;
            if (this.DEBUG) {
                debugGas(`Add return value size fee (${returnFee} to gas used (-> ${totalGas}))`);
            }
        }
        // Check for SpuriousDragon EIP-170 code size limit
        let allowedCodeSize = true;
        if (!result.exceptionError &&
            this.common.gteHardfork(ethereumjs_common_1.Hardfork.SpuriousDragon) &&
            result.returnValue.length > Number(this.common.param('vm', 'maxCodeSize'))) {
            allowedCodeSize = false;
        }
        // If enough gas and allowed code size
        let CodestoreOOG = false;
        if (totalGas <= message.gasLimit && (this.allowUnlimitedContractSize || allowedCodeSize)) {
            if (this.common.isActivatedEIP(3541) && result.returnValue[0] === eof_js_1.EOF.FORMAT) {
                if (!this.common.isActivatedEIP(3540)) {
                    result = { ...result, ...INVALID_BYTECODE_RESULT(message.gasLimit) };
                }
                // Begin EOF1 contract code checks
                // EIP-3540 EOF1 header check
                const eof1CodeAnalysisResults = eof_js_1.EOF.codeAnalysis(result.returnValue);
                if (typeof eof1CodeAnalysisResults?.code === 'undefined') {
                    result = {
                        ...result,
                        ...INVALID_EOF_RESULT(message.gasLimit),
                    };
                }
                else if (this.common.isActivatedEIP(3670)) {
                    // EIP-3670 EOF1 opcode check
                    const codeStart = eof1CodeAnalysisResults.data > 0 ? 10 : 7;
                    // The start of the code section of an EOF1 compliant contract will either be
                    // index 7 (if no data section is present) or index 10 (if a data section is present)
                    // in the bytecode of the contract
                    if (!eof_js_1.EOF.validOpcodes(result.returnValue.subarray(codeStart, codeStart + eof1CodeAnalysisResults.code))) {
                        result = {
                            ...result,
                            ...INVALID_EOF_RESULT(message.gasLimit),
                        };
                    }
                    else {
                        result.executionGasUsed = totalGas;
                    }
                }
            }
            else {
                result.executionGasUsed = totalGas;
            }
        }
        else {
            if (this.common.gteHardfork(ethereumjs_common_1.Hardfork.Homestead)) {
                if (!allowedCodeSize) {
                    if (this.DEBUG) {
                        debug(`Code size exceeds maximum code size (>= SpuriousDragon)`);
                    }
                    result = { ...result, ...CodesizeExceedsMaximumError(message.gasLimit) };
                }
                else {
                    if (this.DEBUG) {
                        debug(`Contract creation: out of gas`);
                    }
                    result = { ...result, ...OOGResult(message.gasLimit) };
                }
            }
            else {
                // we are in Frontier
                if (totalGas - returnFee <= message.gasLimit) {
                    // we cannot pay the code deposit fee (but the deposit code actually did run)
                    if (this.DEBUG) {
                        debug(`Not enough gas to pay the code deposit fee (Frontier)`);
                    }
                    result = { ...result, ...COOGResult(totalGas - returnFee) };
                    CodestoreOOG = true;
                }
                else {
                    if (this.DEBUG) {
                        debug(`Contract creation: out of gas`);
                    }
                    result = { ...result, ...OOGResult(message.gasLimit) };
                }
            }
        }
        // Save code if a new contract was created
        if (!result.exceptionError &&
            result.returnValue !== undefined &&
            result.returnValue.length !== 0) {
            await this.stateManager.putContractCode(message.to, result.returnValue);
            if (this.DEBUG) {
                debug(`Code saved on new contract creation`);
            }
        }
        else if (CodestoreOOG) {
            // This only happens at Frontier. But, let's do a sanity check;
            if (!this.common.gteHardfork(ethereumjs_common_1.Hardfork.Homestead)) {
                // Pre-Homestead behavior; put an empty contract.
                // This contract would be considered "DEAD" in later hard forks.
                // It is thus an unnecessary default item, which we have to save to disk
                // It does change the state root, but it only wastes storage.
                const account = await this.stateManager.getAccount(message.to);
                await this.journal.putAccount(message.to, account ?? new ethereumjs_util_1.Account());
            }
        }
        return {
            createdAddress: message.to,
            execResult: result,
        };
    }
    /**
     * Starts the actual bytecode processing for a CALL or CREATE
     */
    async runInterpreter(message, opts = {}) {
        let contract = await this.stateManager.getAccount(message.to ?? ethereumjs_util_1.Address.zero());
        if (!contract) {
            contract = new ethereumjs_util_1.Account();
        }
        const env = {
            address: message.to ?? ethereumjs_util_1.Address.zero(),
            caller: message.caller ?? ethereumjs_util_1.Address.zero(),
            callData: message.data ?? Uint8Array.from([0]),
            callValue: message.value ?? ethereumjs_util_1.BIGINT_0,
            code: message.code,
            isStatic: message.isStatic ?? false,
            depth: message.depth ?? 0,
            gasPrice: this._tx.gasPrice,
            origin: this._tx.origin ?? message.caller ?? ethereumjs_util_1.Address.zero(),
            block: this._block ?? defaultBlock(),
            contract,
            codeAddress: message.codeAddress,
            gasRefund: message.gasRefund,
            containerCode: message.containerCode,
            blobVersionedHashes: message.blobVersionedHashes ?? [],
            createdAddresses: message.createdAddresses,
        };
        const interpreter = new interpreter_js_1.Interpreter(this, this.stateManager, this.blockchain, env, message.gasLimit, this.journal, this.performanceLogger, this._optsCached.profiler);
        if (message.selfdestruct) {
            interpreter._result.selfdestruct = message.selfdestruct;
        }
        if (message.createdAddresses) {
            interpreter._result.createdAddresses = message.createdAddresses;
        }
        const interpreterRes = await interpreter.run(message.code, opts);
        let result = interpreter._result;
        let gasUsed = message.gasLimit - interpreterRes.runState.gasLeft;
        if (interpreterRes.exceptionError) {
            if (interpreterRes.exceptionError.error !== exceptions_js_1.ERROR.REVERT &&
                interpreterRes.exceptionError.error !== exceptions_js_1.ERROR.INVALID_EOF_FORMAT) {
                gasUsed = message.gasLimit;
            }
            // Clear the result on error
            result = {
                ...result,
                logs: [],
                selfdestruct: new Set(),
                createdAddresses: new Set(),
            };
        }
        return {
            ...result,
            runState: {
                ...interpreterRes.runState,
                ...result,
                ...interpreter._env,
            },
            exceptionError: interpreterRes.exceptionError,
            gas: interpreterRes.runState?.gasLeft,
            executionGasUsed: gasUsed,
            gasRefund: interpreterRes.runState.gasRefund,
            returnValue: result.returnValue ? result.returnValue : new Uint8Array(0),
        };
    }
    /**
     * Executes an EVM message, determining whether it's a call or create
     * based on the `to` address. It checkpoints the state and reverts changes
     * if an exception happens during the message execution.
     */
    async runCall(opts) {
        let message = opts.message;
        let callerAccount;
        if (!message) {
            this._block = opts.block ?? defaultBlock();
            this._tx = {
                gasPrice: opts.gasPrice ?? ethereumjs_util_1.BIGINT_0,
                origin: opts.origin ?? opts.caller ?? ethereumjs_util_1.Address.zero(),
            };
            const caller = opts.caller ?? ethereumjs_util_1.Address.zero();
            const value = opts.value ?? ethereumjs_util_1.BIGINT_0;
            if (opts.skipBalance === true) {
                callerAccount = await this.stateManager.getAccount(caller);
                if (!callerAccount) {
                    callerAccount = new ethereumjs_util_1.Account();
                }
                if (callerAccount.balance < value) {
                    // if skipBalance and balance less than value, set caller balance to `value` to ensure sufficient funds
                    callerAccount.balance = value;
                    await this.journal.putAccount(caller, callerAccount);
                }
            }
            message = new message_js_1.Message({
                caller,
                gasLimit: opts.gasLimit ?? BigInt(0xffffff),
                to: opts.to,
                value,
                data: opts.data,
                code: opts.code,
                depth: opts.depth,
                isCompiled: opts.isCompiled,
                isStatic: opts.isStatic,
                salt: opts.salt,
                selfdestruct: opts.selfdestruct ?? new Set(),
                createdAddresses: opts.createdAddresses ?? new Set(),
                delegatecall: opts.delegatecall,
                blobVersionedHashes: opts.blobVersionedHashes,
            });
        }
        if (message.depth === 0) {
            if (!callerAccount) {
                callerAccount = await this.stateManager.getAccount(message.caller);
            }
            if (!callerAccount) {
                callerAccount = new ethereumjs_util_1.Account();
            }
            callerAccount.nonce++;
            await this.journal.putAccount(message.caller, callerAccount);
            if (this.DEBUG) {
                debug(`Update fromAccount (caller) nonce (-> ${callerAccount.nonce}))`);
            }
        }
        await this._emit('beforeMessage', message);
        if (!message.to && this.common.isActivatedEIP(2929) === true) {
            message.code = message.data;
            this.journal.addWarmedAddress((await this._generateAddress(message)).bytes);
        }
        await this.journal.checkpoint();
        if (this.common.isActivatedEIP(1153))
            this.transientStorage.checkpoint();
        if (this.DEBUG) {
            debug('-'.repeat(100));
            debug(`message checkpoint`);
        }
        let result;
        if (this.DEBUG) {
            const { caller, gasLimit, to, value, delegatecall } = message;
            debug(`New message caller=${caller} gasLimit=${gasLimit} to=${to?.toString() ?? 'none'} value=${value} delegatecall=${delegatecall ? 'yes' : 'no'}`);
        }
        if (message.to) {
            if (this.DEBUG) {
                debug(`Message CALL execution (to: ${message.to})`);
            }
            result = await this._executeCall(message);
        }
        else {
            if (this.DEBUG) {
                debug(`Message CREATE execution (to undefined)`);
            }
            result = await this._executeCreate(message);
        }
        if (this.DEBUG) {
            const { executionGasUsed, exceptionError, returnValue } = result.execResult;
            debug(`Received message execResult: [ gasUsed=${executionGasUsed} exceptionError=${exceptionError ? `'${exceptionError.error}'` : 'none'} returnValue=${(0, ethereumjs_util_1.short)(returnValue)} gasRefund=${result.execResult.gasRefund ?? 0} ]`);
        }
        const err = result.execResult.exceptionError;
        // This clause captures any error which happened during execution
        // If that is the case, then all refunds are forfeited
        // There is one exception: if the CODESTORE_OUT_OF_GAS error is thrown
        // (this only happens the Frontier/Chainstart fork)
        // then the error is dismissed
        if (err && err.error !== exceptions_js_1.ERROR.CODESTORE_OUT_OF_GAS) {
            result.execResult.selfdestruct = new Set();
            result.execResult.createdAddresses = new Set();
            result.execResult.gasRefund = ethereumjs_util_1.BIGINT_0;
        }
        if (err &&
            !(this.common.hardfork() === ethereumjs_common_1.Hardfork.Chainstart && err.error === exceptions_js_1.ERROR.CODESTORE_OUT_OF_GAS)) {
            result.execResult.logs = [];
            await this.journal.revert();
            if (this.common.isActivatedEIP(1153))
                this.transientStorage.revert();
            if (this.DEBUG) {
                debug(`message checkpoint reverted`);
            }
        }
        else {
            await this.journal.commit();
            if (this.common.isActivatedEIP(1153))
                this.transientStorage.commit();
            if (this.DEBUG) {
                debug(`message checkpoint committed`);
            }
        }
        await this._emit('afterMessage', result);
        return result;
    }
    /**
     * Bound to the global VM and therefore
     * shouldn't be used directly from the evm class
     */
    async runCode(opts) {
        this._block = opts.block ?? defaultBlock();
        this._tx = {
            gasPrice: opts.gasPrice ?? ethereumjs_util_1.BIGINT_0,
            origin: opts.origin ?? opts.caller ?? ethereumjs_util_1.Address.zero(),
        };
        const message = new message_js_1.Message({
            code: opts.code,
            data: opts.data,
            gasLimit: opts.gasLimit ?? BigInt(0xffffff),
            to: opts.to ?? ethereumjs_util_1.Address.zero(),
            caller: opts.caller,
            value: opts.value,
            depth: opts.depth,
            selfdestruct: opts.selfdestruct ?? new Set(),
            isStatic: opts.isStatic,
            blobVersionedHashes: opts.blobVersionedHashes,
        });
        return this.runInterpreter(message, { pc: opts.pc });
    }
    /**
     * Returns code for precompile at the given address, or undefined
     * if no such precompile exists.
     */
    getPrecompile(address) {
        return this.precompiles.get((0, ethereumjs_util_1.bytesToUnprefixedHex)(address.bytes));
    }
    /**
     * Executes a precompiled contract with given data and gas limit.
     */
    runPrecompile(code, data, gasLimit) {
        if (typeof code !== 'function') {
            throw new Error('Invalid precompile');
        }
        const opts = {
            data,
            gasLimit,
            common: this.common,
            _EVM: this,
            _debug: this.DEBUG ? debugPrecompiles : undefined,
            stateManager: this.stateManager,
        };
        return code(opts);
    }
    async _loadCode(message) {
        if (!message.code) {
            const precompile = this.getPrecompile(message.codeAddress);
            if (precompile) {
                message.code = precompile;
                message.isCompiled = true;
            }
            else {
                message.containerCode = await this.stateManager.getContractCode(message.codeAddress);
                message.isCompiled = false;
                if (this.common.isActivatedEIP(3540)) {
                    message.code = (0, eof_js_1.getEOFCode)(message.containerCode);
                }
                else {
                    message.code = message.containerCode;
                }
            }
        }
    }
    async _generateAddress(message) {
        let addr;
        if (message.salt) {
            addr = (0, ethereumjs_util_1.generateAddress2)(message.caller.bytes, message.salt, message.code);
        }
        else {
            let acc = await this.stateManager.getAccount(message.caller);
            if (!acc) {
                acc = new ethereumjs_util_1.Account();
            }
            const newNonce = acc.nonce - ethereumjs_util_1.BIGINT_1;
            addr = (0, ethereumjs_util_1.generateAddress)(message.caller.bytes, (0, ethereumjs_util_1.bigIntToBytes)(newNonce));
        }
        return new ethereumjs_util_1.Address(addr);
    }
    async _reduceSenderBalance(account, message) {
        account.balance -= message.value;
        if (account.balance < ethereumjs_util_1.BIGINT_0) {
            throw new exceptions_js_1.EvmError(exceptions_js_1.ERROR.INSUFFICIENT_BALANCE);
        }
        const result = this.journal.putAccount(message.authcallOrigin ?? message.caller, account);
        if (this.DEBUG) {
            debug(`Reduced sender (${message.caller}) balance (-> ${account.balance})`);
        }
        return result;
    }
    async _addToBalance(toAccount, message) {
        const newBalance = toAccount.balance + message.value;
        if (newBalance > ethereumjs_util_1.MAX_INTEGER) {
            throw new exceptions_js_1.EvmError(exceptions_js_1.ERROR.VALUE_OVERFLOW);
        }
        toAccount.balance = newBalance;
        // putAccount as the nonce may have changed for contract creation
        const result = this.journal.putAccount(message.to, toAccount);
        if (this.DEBUG) {
            debug(`Added toAccount (${message.to}) balance (-> ${toAccount.balance})`);
        }
        return result;
    }
    /**
     * Once the interpreter has finished depth 0, a post-message cleanup should be done
     */
    postMessageCleanup() {
        if (this.common.isActivatedEIP(1153))
            this.transientStorage.clear();
    }
    /**
     * This method copies the EVM, current HF and EIP settings
     * and returns a new EVM instance.
     *
     * Note: this is only a shallow copy and both EVM instances
     * will point to the same underlying state DB.
     *
     * @returns EVM
     */
    shallowCopy() {
        const common = this.common.copy();
        common.setHardfork(this.common.hardfork());
        const opts = {
            ...this._optsCached,
            common,
            stateManager: this.stateManager.shallowCopy(),
        };
        opts.stateManager.common = common;
        return new EVM(opts);
    }
    getPerformanceLogs() {
        return this.performanceLogger.getLogs();
    }
    clearPerformanceLogs() {
        this.performanceLogger.clear();
    }
}
exports.EVM = EVM;
EVM.supportedHardforks = [
    ethereumjs_common_1.Hardfork.Chainstart,
    ethereumjs_common_1.Hardfork.Homestead,
    ethereumjs_common_1.Hardfork.Dao,
    ethereumjs_common_1.Hardfork.TangerineWhistle,
    ethereumjs_common_1.Hardfork.SpuriousDragon,
    ethereumjs_common_1.Hardfork.Byzantium,
    ethereumjs_common_1.Hardfork.Constantinople,
    ethereumjs_common_1.Hardfork.Petersburg,
    ethereumjs_common_1.Hardfork.Istanbul,
    ethereumjs_common_1.Hardfork.MuirGlacier,
    ethereumjs_common_1.Hardfork.Berlin,
    ethereumjs_common_1.Hardfork.London,
    ethereumjs_common_1.Hardfork.ArrowGlacier,
    ethereumjs_common_1.Hardfork.GrayGlacier,
    ethereumjs_common_1.Hardfork.MergeForkIdTransition,
    ethereumjs_common_1.Hardfork.Paris,
    ethereumjs_common_1.Hardfork.Shanghai,
    ethereumjs_common_1.Hardfork.Cancun,
    ethereumjs_common_1.Hardfork.Prague,
];
function OOGResult(gasLimit) {
    return {
        returnValue: new Uint8Array(0),
        executionGasUsed: gasLimit,
        exceptionError: new exceptions_js_1.EvmError(exceptions_js_1.ERROR.OUT_OF_GAS),
    };
}
exports.OOGResult = OOGResult;
// CodeDeposit OOG Result
function COOGResult(gasUsedCreateCode) {
    return {
        returnValue: new Uint8Array(0),
        executionGasUsed: gasUsedCreateCode,
        exceptionError: new exceptions_js_1.EvmError(exceptions_js_1.ERROR.CODESTORE_OUT_OF_GAS),
    };
}
exports.COOGResult = COOGResult;
function INVALID_BYTECODE_RESULT(gasLimit) {
    return {
        returnValue: new Uint8Array(0),
        executionGasUsed: gasLimit,
        exceptionError: new exceptions_js_1.EvmError(exceptions_js_1.ERROR.INVALID_BYTECODE_RESULT),
    };
}
exports.INVALID_BYTECODE_RESULT = INVALID_BYTECODE_RESULT;
function INVALID_EOF_RESULT(gasLimit) {
    return {
        returnValue: new Uint8Array(0),
        executionGasUsed: gasLimit,
        exceptionError: new exceptions_js_1.EvmError(exceptions_js_1.ERROR.INVALID_EOF_FORMAT),
    };
}
exports.INVALID_EOF_RESULT = INVALID_EOF_RESULT;
function CodesizeExceedsMaximumError(gasUsed) {
    return {
        returnValue: new Uint8Array(0),
        executionGasUsed: gasUsed,
        exceptionError: new exceptions_js_1.EvmError(exceptions_js_1.ERROR.CODESIZE_EXCEEDS_MAXIMUM),
    };
}
exports.CodesizeExceedsMaximumError = CodesizeExceedsMaximumError;
function EvmErrorResult(error, gasUsed) {
    return {
        returnValue: new Uint8Array(0),
        executionGasUsed: gasUsed,
        exceptionError: error,
    };
}
exports.EvmErrorResult = EvmErrorResult;
function defaultBlock() {
    return {
        header: {
            number: ethereumjs_util_1.BIGINT_0,
            cliqueSigner: () => ethereumjs_util_1.Address.zero(),
            coinbase: ethereumjs_util_1.Address.zero(),
            timestamp: ethereumjs_util_1.BIGINT_0,
            difficulty: ethereumjs_util_1.BIGINT_0,
            prevRandao: (0, ethereumjs_util_1.zeros)(32),
            gasLimit: ethereumjs_util_1.BIGINT_0,
            baseFeePerGas: undefined,
            getBlobGasPrice: () => undefined,
        },
    };
}
exports.defaultBlock = defaultBlock;
