// @nomicfoundation/ethereumjs-evm/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Message = exports.getActivePrecompiles = exports.EVMErrorMessage = exports.EvmError = exports.EVM = exports.EOF = void 0;
const eof_js_1 = require("./eof.js");
Object.defineProperty(exports, "EOF", { enumerable: true, get: function () { return eof_js_1.EOF; } });
const evm_js_1 = require("./evm.js");
Object.defineProperty(exports, "EVM", { enumerable: true, get: function () { return evm_js_1.EVM; } });
const exceptions_js_1 = require("./exceptions.js");
Object.defineProperty(exports, "EVMErrorMessage", { enumerable: true, get: function () { return exceptions_js_1.ERROR; } });
Object.defineProperty(exports, "EvmError", { enumerable: true, get: function () { return exceptions_js_1.EvmError; } });
const message_js_1 = require("./message.js");
Object.defineProperty(exports, "Message", { enumerable: true, get: function () { return message_js_1.Message; } });
const index_js_1 = require("./precompiles/index.js");
Object.defineProperty(exports, "getActivePrecompiles", { enumerable: true, get: function () { return index_js_1.getActivePrecompiles; } });
